shap/explainers/mimic.py

```diff
@@ -0,0 +1,105 @@
+from ..explainers._explainer import Explainer
+
+
+log = logging.getLogger('shap')
+try:
+    import xgboost
+except ImportError:
+    pass
+except:
+    print("xgboost is installed...but failed to load!")
+    pass
+
+class MimicExplainer(Explainer):
+    """Fits a mimic model to the original model and then explains predictions using the mimic model. 
+
+    Tree SHAP allows for very fast SHAP value explainations of flexible gradient boosted decision
+    tree (GBDT) models. Since GBDT models are so flexible we can train them to mimic any black-box
+    model and then using Tree SHAP we can explain them. This won't work well for images, but for
+    any type of problem that GBDTs do reasonable well on, they should also be able to learn how to
+    explain black-box models on the data. This mimic explainer also allows you to use a linear model,
+    but keep in mind that will not do as well at explaining typical non-linear black-box models. In
+    the future we could include other mimic model types given enough demand/help. Finally, we would
+    like to note that this explainer is vaugely inspired by https://arxiv.org/abs/1802.07814 where
+    they learn an explainer that can be applied to any input.
+
+
+    Parameters
+    ----------
+    model : function or iml.Model
+        User supplied function that takes a matrix of samples (# samples x # features) and
+        computes a the output of the model for those samples. The output can be a vector
+        (# samples) or a matrix (# samples x # model outputs).
+
+    data : numpy.array or pandas.DataFrame or iml.DenseData
+        The background dataset to use for integrating out features. To determine the impact
+        of a feature, that feature is set to "missing" and the change in the model output
+        is observed. Since most models aren't designed to handle arbitrary missing data at test
+        time, we simulate "missing" by replacing the feature with the values it takes in the
+        background dataset. So if the background dataset is a simple sample of all zeros, then
+        we would approximate a feature being missing by setting it to zero. For small problems
+        this background datset can be the whole training set, but for larger problems consider
+        using a single reference value or using the kmeans function to summarize the dataset.
+    """
+
+    def __init__(self, model, data, mimic_model="xgboost", mimic_model_params={}, link = IdentityLink(), **kwargs):
+        self.mimic_model_type = mimic_model
+        self.mimic_model_params = mimic_model_params
+
+        # convert incoming inputs to standardized iml objects
+        self.link = convert_to_link(link)
+        self.model = convert_to_model(model)
+        self.keep_index = kwargs.get("keep_index", False)
+        self.data = convert_to_data(data, keep_index=self.keep_index)
+        match_model_to_data(self.model, self.data)
```

Comment:
Do you think maybe the output from `match_model_to_data` should be assigned to `self.model_out`?